stablehlo/dialect/StablehloOps.td

```diff
@@ -2084,6 +2084,44 @@ def StableHLO_CollectivePermuteOp: StableHLO_Op<"collective_permute",
       "::mlir::DenseIntElementsAttr":$source_target_pairs)>];
 }
 
+def StableHLO_CompositeOp : StableHLO_Op<"composite", []> {
+  let summary = "Composite operation";
+  let description = [{
+    Encapsulates an operation made up (composed) of other StableHLO operations,
+    taking `inputs` and producing `results`. The decomposition is specified
+    using the `decomposition` attribute. The composite op can be replaced with
+    its decomposition without changing semantics.
+
+    The `version` and `composite_attributes` enable additional information
+    to be communicated between producers and consumers. `version` defaults to 0
+    and `composite_attributes` defaults to an empty dictionary.
+
+    See:
+    https://github.com/openxla/stablehlo/blob/main/docs/spec.md#composite
+
+    Example:
+    ```mlir
+    %results = stablehlo.composite "my.op" %arg0, %arg1 {
+      decomposition = @my_op,
+      composite_attributes = { my_attribute = "my_value" },
+      version = 1 : i32
+    } : (tensor<f32>, tensor<f32>) -> tensor<f32>
+    ```
+  }];
+
+  let arguments = (ins
+    Variadic<HLO_TensorOrPerAxisQuantizedTensorOrTokenOrTuple>:$inputs, /*composite_i1*/
+    StrAttr:$name, /*composite_i2*/
+    FlatSymbolRefAttr:$decomposition, /*composite_i3*/
+    DefaultValuedOptionalAttr<I32Attr, "0">:$version, /*composite_i4*/
+    DefaultValuedOptionalAttr<DictionaryAttr, "{}">:$composite_attributes /*composite_i5*/
```

Comment:
For reference - I don't think any special default value handling would be needed in VHLO conversions if these were just `DefaultValuedAttr` instead of `DefaultValuedOptionalAttr`. I go back and forth on if I like these. They simplify the pretty-print, may have some memory savings, but otherwise don't provide too much value.

Reply:
Good to know, thanks! I find it kinda weird that both exist TBH. IMO if something has a default value, it is also optional, but I guess "optional" here means something different.